core/src/main/kotlin/org/evomaster/core/problem/graphql/GraphQLUtils.kt

```diff
@@ -165,4 +170,273 @@ object GraphQLUtils {
+                                        history.remove(ob)
+                                    } else {
+                                        history.remove(ob)
+                                    }
+                                }
+                                history.remove(element.tableFieldType)// remove the union type object to the history: the normal case (not cycle)
+                            } else {
+                                history.remove(element.tableFieldType)// remove the union type object to the history:the cycle case
+                            }
+                        } else {
+                            if (element.kindOfTableFieldType.toString().toLowerCase() == "interface") {
+                                history.add(element.tableFieldType)// add interface type object to the history
+                                addEdge(entryPoint, element.tableFieldType, element.tableField, graph)//from a node to the interface type object
+                                if (history.count { it == element.tableFieldType } <= 1) {// the interface type object already treated
+                                    constructGraph(state, element.tableFieldType, "", graph, history, objectFieldsHistory)//construct the base fields of the interface
+                                    element.interfaceTypes.forEach { ob ->
+                                        history.add(ob)//add the interface objects to the history
+                                        addEdge(element.tableFieldType, ob, interfaceTag, graph) // from the interface type object to its objects
+                                        if (history.count { it == ob } <= 1) {
+                                            constructGraph(state, ob, "", graph, history, objectFieldsHistory)
+                                            history.remove(ob)
+                                        } else {
+                                            history.remove(ob)
+                                        }
+                                    }
+                                    history.remove(element.tableFieldType)// remove the interface type object to the history:the normal case (not cycle)
+                                } else {
+                                    history.remove(element.tableFieldType)// remove the interface type object to the history:the cycle case
+                                }
+                            }
+                        }
+                    }
+
+                }
+            }
+        }
+        return graph
+    }
+
+    private fun addNode(node: String, graph: MutableMap<String, GraphInfo>): GraphInfo? {
+        return graph.putIfAbsent(node, GraphInfo())
+    }
+
+    private fun addEdge(source: String, destination: String, fieldName: String, graph: MutableMap<String, GraphInfo>): Boolean? {
+        addNode(source, graph)
+        addNode(destination, graph)
+        graph[source]?.edges?.add(destination)
```

Comment:
with `addNode(source, graph)`, is it not guaranteed that `graph[source]` is not null? (and if it is, then it is a bug in `addNode`). so why the `?`?

Reply:
-Because of the function: `putIfAbsent()` in `addNode(source, graph)` :

`Returns: the previous value associated with the specified key, or null if there was no mapping for the key. (A null return can also indicate that the map previously associated null with the key, if the implementation supports null values.)`